pyface/data_view/abstract_data_model.py

```diff
@@ -0,0 +1,272 @@
+
+""" Provides an AbstractDataModel ABC for Pyface data models.
+
+This module provides an ABC for all data view data models.  This specifies
+the API that the data view widgets expect, and which the underlying
+data is adapted to by the concrete implementations.  Data models are intended
+to be toolkit-independent, and be able to adapt any approximately tabular or
+nested data structure to what the data view system expects.
+"""
+from abc import abstractmethod
+
+from traits.api import ABCHasStrictTraits, Event, Instance
+
+from .index_manager import AbstractIndexManager
+
+
+class DataViewSetError(ValueError):
+    """ An exception raised when setting a value fails. """
+    pass
+
+
+class AbstractDataModel(ABCHasStrictTraits):
+    """ Abstract base class for Pyface data models.
+
+    The data model API is intended to provide a common API for hierarchical
+    and tabular data.  This class is concerned with the structure, type and
+    values provided by the data, but not with how the data is presented.
+
+    Row and column indices are represented by sequences (usually lists) of
+    integers, specifying the index at each level of the heirarchy.  The root
+    row and column are represented by empty lists.
+
+    Subclasses need to implement the ``get_column_count``,
+    ``can_have_children`` and ``get_row_count`` methods to return the number
+    of columns in a particular row, as well as the hierarchical structure of
+    the rows.  Appropriate observers should be set up on the underlaying data
+    so that the ``structure_changed`` event is fired when the values returned
+    by these methods would change.
+
+    Subclasses also have to implement the ``get_value`` and ``get_value_type``
+    methods.  These expect a row and column index, with root values treated
+    specially: the root row corresponds to the values which will be displayed
+    in the column headers of the view, and the root column corresponds to the
+    values which will be displayed in the row headers of the view.
+    The ``get_value`` returns an arbitrary Python object corresponding to the
+    cell being viewed, and the ``get_value_type`` should return an instance of
+    an ``AbstractValueType`` that adapts the raw value to the data channels
+    that the data view expects (eg. text, color, icons, editable value, etc.).
+    Implementations should ensure that the ``values_changed`` event fires
+    whenever the data, or the way the data is presented, is updated.
+
+
+    If the data is to be editable then the subclass should override the
+    ``set_value`` method.  It should attempt to change the underlying data as a
+    side-effect, and return True on success and False on failure (for example,
+    setting an invalid value).  If the underlying data structure cannot be
+    listened to internally (such as a numpy array or Pandas data frame), this
+    method should also fire the values changed event with appropriate values.
```

Comment:
No, because this will have people adding code to fire events in `set_value` when they should be writing observers.

The correct admonition for writers of models is that they need to ensure that the events are fired _whenever_ the data changes (either through set_value or externally).

Reply:
> The correct admonition for writers of models is that they need to ensure that the events are fired whenever the data changes (either through set_value or externally).

Agreed. Could we remove the last sentence starting with "If the underlying data structure cannot be listened to..." then? It is confusing and it can already be implied by the more correct and clear rule (which is already in the paragraph above this one, which reads "Implementations should ensure that the ``values_changed`` event fires")

Reply:
Thanks. Resolving...